packages/zoe/state.js

```diff
@@ -1,190 +1,279 @@
+  const table = harden({
+    validate: validateFn,
+    create: (record, handle = harden({})) => {
+      record = harden({
+        ...record,
+        handle, // reliably add the handle to the record
       });
+      table.validate(record);
+      handleToRecord.init(handle, record);
+      return handle;
     },
-    recordOffer: (offerHandle, offerRules, extents, assays, result) => {
-      const { payoutRules, exit } = offerRules;
-      offerHandleToExtents.init(offerHandle, extents);
-      offerHandleToAssays.init(offerHandle, assays);
-      offerHandleToPayoutRules.init(offerHandle, payoutRules);
-      offerHandleToExitRule.init(offerHandle, exit);
-      offerHandleToResult.init(offerHandle, result);
-      activeOffers.add(offerHandle);
-    },
-    recordUsedInInstance: (instanceHandle, offerHandle) =>
-      offerHandleToInstanceHandle.init(offerHandle, instanceHandle),
-    getInstanceHandleForOfferHandle: offerHandle => {
-      if (offerHandleToInstanceHandle.has(offerHandle)) {
-        return offerHandleToInstanceHandle.get(offerHandle);
-      }
-      return undefined;
-    },
-    setExtentsFor: (offerHandles, reallocation) =>
-      offerHandles.map((offerHandle, i) =>
-        offerHandleToExtents.set(offerHandle, reallocation[i]),
-      ),
-    getResultsFor: offerHandles =>
-      offerHandles.map(offerHandle => offerHandleToResult.get(offerHandle)),
-    removeOffers: offerHandles => {
-      // has-side-effects
-      // eslint-disable-next-line array-callback-return
-      offerHandles.map(offerHandle => {
-        offerHandleToExtents.delete(offerHandle);
-        offerHandleToAssays.delete(offerHandle);
-        offerHandleToPayoutRules.delete(offerHandle);
-        offerHandleToExitRule.delete(offerHandle);
-        offerHandleToResult.delete(offerHandle);
-        if (offerHandleToInstanceHandle.has(offerHandle)) {
-          offerHandleToInstanceHandle.delete(offerHandle);
-        }
+    get: handleToRecord.get,
+    has: handleToRecord.has,
+    delete: handleToRecord.delete,
+    update: (handle, partialRecord) => {
+      const record = handleToRecord.get(handle);
+      const updatedRecord = harden({
+        ...record,
+        ...partialRecord,
       });
+      table.validate(updatedRecord);
+      handleToRecord.set(handle, updatedRecord);
+      return record;
     },
-    setOffersAsInactive: offerHandles => {
-      offerHandles.map(offerHandle => activeOffers.delete(offerHandle));
-    },
+    // eslint-disable-next-line no-use-before-define
   });
-  return {
-    adminState,
-    readOnlyState,
+
+  const customMethodsTable = harden({
+    ...makeCustomMethodsFn(table),
+    ...table,
+  });
+  return customMethodsTable;
+};
+
+const validateProperties = (expectedProperties, obj) => {
+  // add handle to expected properties
+  expectedProperties.push('handle');
+  const actualProperties = Object.getOwnPropertyNames(obj);
+  // Sorts in-place
+  expectedProperties.sort();
```

Comment:
No problem! I curried it so that the sorting only happens once. I don't think it makes sense for the user to pre-sort the expected properties though - I could see someone easily forgetting and it causing errors. Would be possible to not harden this parameter, especially since it is only used by other state.js code?

Reply:
Ah, gotcha. I understand now. That works for me